ql/time/daycounters/actualactual.cpp

```diff
@@ -40,111 +44,120 @@ namespace QuantLib {
         }
     }
 
+	int ActualActual::ISMA_Impl::findCouponsPerYear(Date refStart, Date refEnd) const {
+		// This will only work for day counts longer than 15 days.
+		Integer months = Integer(0.5 + 12 * Real(dayCount(refStart, refEnd))/365.0);
+		return (int)round(12.0 / Real(months));
+
+	}
+
+	Time ActualActual::ISMA_Impl::yearFractionWithReferenceDates(
+		const Date& d1,
+		const Date& d2,
+		const Date& d3,
+		const Date& d4
+	) const {
+		QL_REQUIRE(
+			d1 <= d2, 
+			"This function is only correct if d1 <= d2 \n d1: " << d1 << " d2: " << d2
+		);
+		//
+		Real referenceDayCount = Real(dayCount(d3, d4));
+		//guess how many coupon periods per year:
+		int couponsPerYear;
+		if (referenceDayCount < 16) {
+			couponsPerYear = 1;
+			referenceDayCount = dayCount(d1, d1 + 1 * Years);
+		}
+		else {
+			couponsPerYear = findCouponsPerYear(d3, d4);
+		}
+		return Real(dayCount(d1, d2)) / (referenceDayCount*couponsPerYear);
```

Comment:
Yes. I think it handles long and short first coupons. It does it because it explictly creates all necessary quasi-periods. It does that through the method: 
getListOfPeriodDatesIncludingQuasiPayments and at that point you can always break up a long period into two. 

I think this is superior to the former approach because without access to the schedule the method had to guess in strange ways, and those guesses are not going to cover the full possibilities. For example, there are plenty of cases where you need to know if the bond uses the End of Month adjustments to work out where the quasi period should be.

I am not sure if this will work for case (e) as it currently stands, as I did not know that this could happen, but we would only need to make adjustments to the method which adds the quasiPeriods to add an extra date at the end and then it would work correctly.

I also feel the logic of explicitly creating the quasi periods so that all these cases become exactly like the base case is more elegant.

Reply:
Yes&mdash;I was thinking of the case in which the version with four dates is called explicitly without a schedule, though.

Anyway: I agree with you that the correct case is with the schedule, and the other should be phased out at some point.  To facilitate this, and in the meantime preserve the current behavior without schedule, I was thinking of using your implementation when a schedule is available (in fact, it could even be simplified if it can assume that a schedule always exist) and to keep the old one when a schedule is not given.  The implementation can be chosen by the constructor.

I'll work a bit on this and open a pull request to you.